src/herring/model.hpp

```diff
@@ -0,0 +1,275 @@
+          case element_op::signed_square:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::copysign(elem * elem, elem);
+            };
+          case element_op::hinge:
+            return [](sum_elem_type elem, float* out) {
+              *out = elem > sum_elem_type{} ? sum_elem_type{1} : sum_elem_type{0};
+            };
+          case element_op::sigmoid:
+            return [constant](sum_elem_type elem, float* out) {
+              *out = sum_elem_type{1} / (sum_elem_type{1} + std::exp(-constant * elem));
+            };
+          case element_op::exponential:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::exp(elem);
+            };
+          case element_op::exponential_standard_ratio:
+            return [constant](sum_elem_type elem, float* out) {
+              *out = std::exp(-elem / constant);
+            };
+          case element_op::logarithm_one_plus_exp:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::log1p(std::exp(elem));
+            };
+          default:
+            return [](sum_elem_type elem, float* out) {
+              *out = elem;
+            };
+        }
+      }();
+    }
+
+   private:
+    std::function<void(sum_elem_type, float*)> postprocess_element;
+
+    auto precompute_missing(float const* input, std::size_t num_row) const {
+      auto result = false;
+      if (num_row > row_has_missing.size()) {
+        row_has_missing.resize(num_row);
+      }
+      for (auto row_index = std::size_t{}; row_index < num_row; ++row_index) {
+        row_has_missing[row_index] = std::transform_reduce(
+          input + row_index * num_feature,
+          input + (row_index + 1) * num_feature,
+          false,
+          std::logical_or<>(),
+          [](auto val) { return std::isnan(val); }
+        );
+        result = result || row_has_missing[row_index];
+      }
+      return result;
+    }
+
+    void apply_postprocessing(
+        std::vector<sum_elem_type> const& grove_sum,
+        float* output,
+        std::size_t num_row,
+        std::size_t num_grove,
+        thread_count<int> nthread) const {
+
+      if (row_postproc != row_op::max_index) {
+#pragma omp parallel for num_threads(static_cast<int>(nthread))
+        for (auto row_index = std::size_t{}; row_index < num_row; ++row_index) {
+          auto const grove_output_index = row_index * num_class * num_grove;
+          for (auto class_index = std::size_t{}; class_index < num_class; ++class_index) {
+            auto const class_output_index = grove_output_index + class_index * num_grove;
+            auto const grove_sum_begin = std::begin(grove_sum);
+            postprocess_element(std::reduce(
+              grove_sum_begin + class_output_index,
+              grove_sum_begin + class_output_index + num_grove,
+              bias * average_factor
+            ) / average_factor, output + row_index * num_class + class_index);
```

Comment:
The handling of `bias` can be tricky in present of `average_factor`. Have you tested for correctness with XGBoost and LightGBM models? (For LightGBM models, test both `rf` and `gbdt` types)

Reply:
The logic does look correct to my eyes. If we're pressed on time, we could create a GH issue to follow up with this testing.

Reply:
No, I think this is certainly worth testing. I'll try to set up a lightgbm test a little later.

Reply:
Great! Take a look at https://github.com/dmlc/treelite/blob/4cc4f7e61abb624c65bdb7ff83972ca4b11ff10d/tests/python/test_gtil.py#L290-L311 for an example of testing LightGBM with two boosting types.